src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/GCHandle.cs

```diff
@@ -191,5 +194,24 @@ private static void ThrowIfInvalid(IntPtr handle)
                 ThrowHelper.ThrowInvalidOperationException_HandleIsNotInitialized();
             }
         }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        internal static unsafe void CheckUninitialized(IntPtr handle)
+        {
+            // Check if the handle was never initialized or was freed.
+            // Throws NRE with minimal overhead, to avoid access violation from unmanaged code.
+            // Invalid handle is unsupported and will cause AV as expected.
+#if MONO
+            // Mono doesn't handle reading null pointer as NRE.
+            // Throw a NRE manually.
+            if (handle == 0)
+            {
+                throw new NullReferenceException();
+            }
+#else
```

Comment:
Can we make this `#elif RELEASE`? I know the JIT should be able to drop this extra read, but it feels like there's no need to unnecessarily make its life more difficult and complicate things when we know this whole statement is unnecessary anyway?

Reply:
I think you meant `!RELEASE`. 

This statement is necessary except for the one call from the getter. The ifdef would have to be in the getter.

Reply:
Sorry, yes, I meant `!RELEASE`. And I was indeed thinking of the getter. I just meant it seems like we should be able to skip this preemptive read in those cases. Adding an ifdef there around this call seems perfect 😄

Reply:
I'm only sure that coreclr (and probably nativeaot) release are fine. The behavior of Mono (especially interpreter) seems to be different.

Reply:
Yeah that's why I was suggesting to add the condition as an `elif` after the check for `MONO`. If we want to move this condition in the callers, we'd likely want to check for both `!MONO` and `!RELEASE` there.

Reply:
The method is used to check the handle for getters and setters. The check can be skipped for getters on release builds on CoreCLR/NAOT. This check cannot be skipped for setters. We want to throw proper `NullReferenceException` instead of fail fasting with runtime fatal error in case somebody calls target setter on a null handle.

Reply:
Yup, no I agree with that, I was only referring to the getters, not to all callers in general. I also would expect getters to be in hot paths way more often than setters anyway, so that seems completely fine 🙂

Reply:
@huoyaoyuan Do you plan to do anything about this feedback? The change looks good to me otherwise.

Reply:
As my previous comment said, the duplicated read can be perfectly eliminated by JIT in coreclr/Release/getter. In most other cases the null check is required, so I prefer to keep the code as-is make it simpler.